File: server/schemas/game/games.py

from pydantic import BaseModel
from typing import Any, TYPE_CHECKING
from schemas.game.enums import MoveType, ActionType, EntityZoneType, ZoneType
from schemas.db.cards import CardSchemas
if TYPE_CHECKING:
    from modules.card import Card
    from modules.effect import Effect
    from modules.player import Player


class Entity(BaseModel):
    zone: EntityZoneType
    index: int
    opponent: bool
    class Config:
        use_enum_values = True

class ConditionInfo(BaseModel):
    player: 'Player'
    opponent: 'Player'
    trigger_cards: 'Card'
    

class CardInfo(CardSchemas):
    zone: ZoneType
    index: int
    opponent: bool
    side_effects: list[Any] = []
    effects: list[int]
    class Config:
        use_enum_values = True
    
class MoveEffect(BaseModel):
    move_id: int
    entity: Entity
    effect_id: int      
    select: bool
    targets: list[Entity]
    tmp: Any

class MoveAttack(BaseModel):
    move_id: int
    entity: Entity
    select: bool
    targets: list[Entity]
    tmp: Any
    
class Move(BaseModel):
    effects: list[MoveEffect] = []
    attact: list[MoveAttack] = []
    end: bool
    
    
class MoveReturn(BaseModel):
    move_type: MoveType
    move_id: int
    target: list[int]
    tmp: Any
    class Config:
        use_enum_values = True

class PlayerInfo(BaseModel):
    cost: int
    health: int
    hands: list[CardInfo]
    fields: dict[int, CardInfo] = {}
    graves: list[CardInfo]
    decks: int

class GameStat(BaseModel):
    Player: PlayerInfo
    opponent: PlayerInfo
    trun: int
    is_player_turn: bool
    class Config:
        use_enum_values = True


class Action(BaseModel):
    action_type: ActionType
    subject: Entity  
    object: Entity
    subject_state: CardInfo | PlayerInfo
    object_state: CardInfo | PlayerInfo

File: client/src/types/games.ts
import { Card } from "./models";

// Enum 타입 정의
export type MessageType = "ping" | "text" |"game_info" |"action" | "move"

export type ZoneType = "hands" | "fields" | "graves" | "decks";

export type MoveType = "effect" | "attack" | "end";

export type ActionType =
  | "move"
  | "card_state"
  | "side_effect"
  | "cost"
  | "attack"
  | "destroy"
  | "damage"
  | "effect";

export type EntityZoneType = "hands" | "fields" | "graves" | "decks" | "player";

export interface MessageModel {
  type: MessageType;
  data: any;
}

export interface Entity {
  zone: EntityZoneType;
  index: number;
  opponent: boolean;
}

export interface CardInfo extends Card {
  side_effects: number[];
}

export interface Move {
  move_type: MoveType;
  entity: Entity;
  select: boolean;
  targets: Entity[];
  effect_id: number | null;
}

export interface MoveReturn {
  move_type: MoveType;
  move_id: number;
  target: number[];
}

export interface PlayerInfo {
  cost: number;
  health: number;
  side_effects: number[];
  hands: CardInfo[];
  fields: Record<number, CardInfo | null>;
  graves: CardInfo[];
  decks: number;
}

export interface Opponent {
  cost: number;
  health: number;
  side_effects: number[];
}

export interface GameInfo {
  player: PlayerInfo;
  opponent: PlayerInfo;
  turn: number;
  is_player_turn: boolean;
  side_effects: number[];
}

export interface GameStat {
  turn: number;
  is_player_turn: boolean;
  side_effects: number[];
}

export interface Player {
  cost: number;
  health: number;
  side_effects: number[];
}
